snmpwalk/check.py

```diff
@@ -0,0 +1,214 @@
+    pysnmp in our use-cases. It is not a 100% reimplementation. It only supports
+    the functionality that we currently use. Other differences include:
+        * regex matching for dynamic tagging w/additional_tags
+        * tags using "enum" values emit the human-readable form instead of the
+          integer.
+    '''
+    DEFAULT_SNMPWALK_PATH = '/usr/bin/snmpwalk'
+    DEFAULT_RETRIES = 2
+    DEFAULT_TIMEOUT = 1
+    COUNTER_TYPES = frozenset(('Counter32', 'Counter64', 'ZeroBasedCounter64'))
+    GAUGE_TYPES = frozenset(('Gauge32', 'Unsigned32', 'CounterBasedGauge64',
+                             'INTEGER', 'Integer32'))
+    SC_NAME = '{}.can_check'.format(SOURCE_TYPE_NAME)
+
+    # regex for parsing the output of snmp walk
+    output_re = re.compile(r'([\w\-]+)::(?P<symbol>\w+)\.(?P<index>\d+) = '
+                        r'(?P<type>\w+): (?P<value>.*)$')
+
+    def __init__(self, name, init_config, agentConfig, instances=None):
+        self.binary = None
+        self.expected_bin = init_config.get('binary', self.DEFAULT_SNMPWALK_PATH)
+        if os.path.isfile(self.expected_bin):
+            self.binary = self.expected_bin
+
+        self.mib_dirs = init_config.get('mibs_folder')
+
+        for instance in instances:
+            # if we don't have a name add one, but mark skip_event so that we
+            # don't emit the event
+            if 'name' not in instance:
+                instance['name'] = self._get_instance_addr(instance)
+            instance['skip_event'] = True
+
+        NetworkCheck.__init__(self, name, init_config, agentConfig, instances)
+
+    def _get_instance_addr(self, instance):
+        host = instance.get('host', None)
+        ip = instance.get('ip_address', None)
+        port = instance.get('port', None)
+        if host and port:
+            key = "{host}:{port}".format(host=host, port=port)
+        elif ip and port:
+            key = "{host}:{port}".format(host=ip, port=port)
+        elif host:
+            key = host
+        elif ip:
+            key = ip
+
+        return key
+
+    def _check(self, instance):
+        if not self.binary:
+            raise BinaryUnavailable("Cannot find executable: {}".format(self.expected_bin))
+
+        ip_address = self._get_instance_addr(instance)
+        metrics = instance.get('metrics', [])
+        community_string = instance.get('community_string', 'public')
+        timeout = int(instance.get('timeout', self.DEFAULT_TIMEOUT))
+        retries = int(instance.get('retries', self.DEFAULT_RETRIES))
+
+        hostname = instance.get('metric_host', None)
+
+        # Build up our dataset
+        data = defaultdict(dict)
+        types = {}
+        for metric in metrics:
+            mib = metric['MIB']
+            table = metric['table']
+            cmd = [self.binary, '-c{}'.format(community_string),
+                   '-v2c', '-t', str(timeout), '-r', str(retries)]
```

Comment:
@jescholl I can definitely see that happen. Enabling the actual `snmpwalk` to make the v3 request should be straight-forward. Wondering how much work the parser might require.